File: app/src/main/java/com/glookogluroo/bridge/glooko/GlookoClient.kt
package com.glookogluroo.bridge.glooko

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.FormBody
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.UUID
import java.util.concurrent.TimeUnit

class GlookoClient(
    private val email: String,
    private val password: String,
    private val sessionTimeoutMinutes: Long = 55,
    private val baseUrl: String = GLOOKO_BASE_URL,
    private val httpClient: OkHttpClient = defaultHttpClient(),
) {
    private var patientId: String? = null
    private var apiBase: String? = null
    private var dashboardOrigin: String? = null
    private var authMethod: String? = null
    private var lastAuthTime: Instant? = null

    init {
        authenticate()
    }

    fun ensureAuthenticated() {
        if (lastAuthTime == null) {
            authenticate()
            return
        }
        val elapsedMinutes = (Instant.now().epochSecond - lastAuthTime!!.epochSecond) / 60
        if (elapsedMinutes >= sessionTimeoutMinutes) {
            authenticate()
        }
    }

    fun authenticate() {
        patientId = null
        apiBase = null
        dashboardOrigin = null
        authMethod = null

        val regional = discoverRegionalHosts()
        dashboardOrigin = regional.dashboardOrigin
        apiBase = regional.apiBase

        val jsonLoginError = tryJsonApiLogin(regional)
        if (patientId != null) {
            authMethod = "json-api"
            lastAuthTime = Instant.now()
            return
        }

        tryWebFormLogin(regional, jsonLoginError)
        authMethod = "web-form"
        lastAuthTime = Instant.now()
    }

    private fun discoverRegionalHosts(): RegionalHosts {
        val loginUrl = "$baseUrl$GLOOKO_LOGIN_PATH?id=login_form&locale=en-GB"
        val initialRequest = Request.Builder().url(loginUrl).get().build()
        val initialResponse = httpClient.newCall(initialRequest).execute()
        initialResponse.close()

        var regionalLoginUrl = initialResponse.header("Location") ?: loginUrl
        if (!regionalLoginUrl.startsWith("http")) {
            regionalLoginUrl = resolveUrl(baseUrl, regionalLoginUrl)
        }

        val regionalHttpUrl = regionalLoginUrl.toHttpUrlOrNull()
            ?: throw GlookoAuthError("Could not resolve regional login URL: $regionalLoginUrl")
        val dashboardOrigin = webOrigin(regionalHttpUrl)
        val apiBase = deriveApiBase(regionalHttpUrl)

        return RegionalHosts(
            regionalLoginUrl = regionalLoginUrl,
            dashboardOrigin = dashboardOrigin,
            apiBase = apiBase,
        )
    }

    private fun tryJsonApiLogin(regional: RegionalHosts): String? {
        val body = buildJsonObject {
            putJsonObject("userLogin") {
                put("email", email)
                put("password", password)
            }
            putJsonObject("deviceInformation") {
                put("applicationType", "logbook")
                put("os", "android")
                put("osVersion", "14")
                put("device", "Android")
                put("deviceManufacturer", "Google")
                put("deviceModel", "Mobile")
                put("serialNumber", DEVICE_ID)
                put("clinicalResearch", false)
                put("deviceId", DEVICE_ID)
                put("applicationVersion", "6.1.3")
                put("buildNumber", "0")
                put("gitHash", "g4fbed2011b")
            }
        }.toString()

        val url = "${regional.apiBase}/api/v2/users/sign_in"
        val request = Request.Builder()
            .url(url)
            .post(body.toRequestBody(JSON_MEDIA_TYPE))
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .header("Origin", regional.dashboardOrigin)
            .header("Referer", "${regional.dashboardOrigin}/")
            .build()

        val response = httpClient.newCall(request).execute()
        response.use {
            val responseBody = it.body?.string().orEmpty()
            if (!it.isSuccessful) {
                return "HTTP ${it.code}: ${responseBody.take(200)}"
            }
            val glookoCode = extractGlookoCodeFromLoginResponse(responseBody)
            if (glookoCode.isNullOrBlank()) {
                return "HTTP ${it.code} but no glookoCode in response: ${responseBody.take(200)}"
            }
            patientId = glookoCode
            return null
        }
    }

    private fun tryWebFormLogin(regional: RegionalHosts, jsonLoginError: String?) {
        val regionalRequest = Request.Builder().url(regional.regionalLoginUrl).get().build()
        val regionalResponse = httpClient.newCall(regionalRequest).execute()
        regionalResponse.use { response ->
            if (!response.isSuccessful) {
                throw GlookoAuthError("Failed to load regional login page: HTTP ${response.code}")
            }
            val html = response.body?.string().orEmpty()
            val csrfToken = CSRF_TOKEN_REGEX.find(html)?.groupValues?.get(1)
                ?: throw GlookoAuthError("Could not extract CSRF token from login page")

            val loginBody = FormBody.Builder()
                .add("utf8", "\u2713")
                .add("authenticity_token", csrfToken)
                .add("user[email]", email)
                .add("user[password]", password)
                .add("commit", "Log In")
                .build()

            val authRequest = Request.Builder()
                .url(regional.regionalLoginUrl)
                .post(loginBody)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .header("Referer", regional.regionalLoginUrl)
                .build()

            val authResponse = httpClient.newCall(authRequest).execute()
            var (dashboardHtml, finalUrl) = followRedirectsToDashboard(authResponse)

            var extractedPatientId = extractPatientId(dashboardHtml)
            if (extractedPatientId == null && !finalUrl.contains("users/sign_in")) {
                val dashboardFetch = fetchUrl("${regional.dashboardOrigin}/")
                dashboardHtml = dashboardFetch.first
                finalUrl = dashboardFetch.second
                extractedPatientId = extractPatientId(dashboardHtml)
            }

            patientId = extractedPatientId
            if (patientId == null) {
                throw GlookoAuthError(
                    buildAuthFailureMessage(
                        dashboardHtml = dashboardHtml,
                        finalUrl = finalUrl,
                        jsonLoginError = jsonLoginError,
                    ),
                )
            }

            apiBase = extractApiBase(dashboardHtml, finalUrl, regional.apiBase)
            dashboardOrigin = regional.dashboardOrigin
        }
    }

    private fun buildAuthFailureMessage(
        dashboardHtml: String,
        finalUrl: String,
        jsonLoginError: String?,
    ): String = buildString {
        append("Could not complete Glooko sign-in")
        if (dashboardHtml.contains("users/sign_in")) {
            append(" (redirected back to login — session may not have been accepted)")
        } else {
            append(" (signed in but patient ID was not found in dashboard)")
        }
        jsonLoginError?.let { append("\nAPI login: $it") }
        append("\nFinal URL: $finalUrl")
        append("\nTry opening the Glooko app once, then retry.")
    }

    private fun followRedirectsToDashboard(initialResponse: okhttp3.Response): Pair<String, String> {
        var response = initialResponse
        var hops = 0
        try {
            while (response.code in REDIRECT_CODES && hops < MAX_REDIRECTS) {
                val location = response.header("Location")
                    ?: throw GlookoAuthError("Login redirect missing Location header")
                val referer = response.request.url.toString()
                val nextUrl = resolveUrl(referer, location)
                response.close()
                val nextRequest = Request.Builder()
                    .url(nextUrl)
                    .get()
                    .header("Referer", referer)
                    .build()
                response = httpClient.newCall(nextRequest).execute()
                hops++
            }

            if (!response.isSuccessful) {
                throw GlookoAuthError("Login failed: HTTP ${response.code}")
            }

            val html = response.body?.string().orEmpty()
            val finalUrl = response.request.url.toString()
            return html to finalUrl
        } finally {
            response.close()
        }
    }

    private fun fetchUrl(url: String): Pair<String, String> {
        val request = Request.Builder().url(url).get().build()
        val response = httpClient.newCall(request).execute()
        return response.use {
            if (!it.isSuccessful) {
                throw GlookoAuthError("Failed to load $url: HTTP ${it.code}")
            }
            it.body?.string().orEmpty() to it.request.url.toString()
        }
    }

    private fun resolveUrl(currentUrl: String, location: String): String {
        if (location.startsWith("http")) return location
        val base = currentUrl.toHttpUrlOrNull()
            ?: throw GlookoAuthError("Could not resolve redirect URL")
        return base.resolve(location)?.toString()
            ?: throw GlookoAuthError("Could not resolve redirect URL: $location")
    }

    private fun extractPatientId(html: String): String? {
        PATIENT_ID_PATTERNS.forEach { pattern ->
            pattern.find(html)?.groupValues?.get(1)?.let { return it }
        }
        return null
    }

    private fun extractGlookoCodeFromLoginResponse(responseBody: String): String? {
        val root = runCatching { GlookoParser.parseJsonObject(responseBody) }.getOrNull() ?: return null
        return jsonStringAt(root, "userLogin", "glookoCode")
            ?: jsonStringAt(root, "user", "userLogin", "glookoCode")
            ?: jsonStringAt(root, "glookoCode")
    }

    private fun jsonStringAt(root: JsonObject, vararg path: String): String? {
        var current: kotlinx.serialization.json.JsonElement = root
        for (key in path) {
            val next = (current as? JsonObject)?.get(key) ?: return null
            current = next
        }
        return runCatching { current.jsonPrimitive.content }.getOrNull()
    }

    private fun extractApiBase(html: String, finalUrl: String, fallbackApiBase: String): String {
        API_URL_REGEX.find(html)?.groupValues?.get(1)?.let { return it }
        val httpUrl = finalUrl.toHttpUrlOrNull()
        if (httpUrl != null) {
            return deriveApiBase(httpUrl)
        }
        return fallbackApiBase
    }

    private fun deriveApiBase(webUrl: okhttp3.HttpUrl): String {
        val apiHost = if (webUrl.host.contains("my.glooko")) {
            webUrl.host.replace("my.glooko", "api.glooko")
        } else {
            webUrl.host
        }
        return webOrigin(webUrl.newBuilder().host(apiHost).build())
    }

    private fun webOrigin(url: okhttp3.HttpUrl): String {
        return url.newBuilder()
            .encodedPath("/")
            .query(null)
            .fragment(null)
            .build()
            .toString()
            .trimEnd('/')
    }

    fun getGraphData(startDate: Instant, endDate: Instant): JsonObject? {
        ensureAuthenticated()
        val start = formatIso(startDate)
        val end = formatIso(endDate)
        val query = buildString {
            append("patient=$patientId")
            append("&startDate=$start")
            append("&endDate=$end")
            append("&locale=en-GB")
            append("&insulinTooltips=true")
            append("&filterBgReadings=true")
            append("&splitByDay=false")
            append("&series[]=deliveredBolus")
        }
        val url = "$apiBase/api/v3/graph/data?$query"
        return apiRequest(url, preserveUrl = true)
    }

    fun getStatistics(startDate: Instant, endDate: Instant): JsonObject? {
        ensureAuthenticated()
        val start = formatIso(startDate)
        val end = formatIso(endDate)
        val query = buildString {
            append("patient=$patientId")
            append("&startDate=$start")
            append("&endDate=$end")
            append("&egv=false")
            append("&includeInsulin=true")
            append("&includeExercise=true")
            append("&dow=monday,tuesday,wednesday,thursday,friday,saturday,sunday")
            append("&includePumpModes=true")
        }
        val url = "$apiBase/api/v3/graph/statistics/overall?$query"
        return apiRequest(url)
    }

    fun getDeviceSettings(): JsonObject? {
        ensureAuthenticated()
        val url = "$apiBase/api/v3/devices_and_settings?patient=$patientId"
        return apiRequest(url)
    }

    fun testConnection(): Result<GlookoTestResult> = runCatching {
        ensureAuthenticated()
        val url = "$apiBase/api/v3/devices_and_settings?patient=$patientId"
        val response = executeRequest(url)
        if (!response.isSuccessful) {
            throw GlookoApiError(url, response.code, response.bodyPreview)
        }
        if (response.body.isBlank()) {
            throw GlookoApiError(url, response.code, "(empty JSON body)")
        }
        val deviceData = GlookoParser.parseJsonObject(response.body)
        val devices = GlookoParser.parseDevices(deviceData)
        GlookoTestResult(
            patientId = patientId ?: "unknown",
            apiBase = apiBase ?: "unknown",
            dashboardOrigin = dashboardOrigin ?: "unknown",
            deviceCount = devices.size,
        )
    }

    fun sessionSummary(): String? {
        if (patientId == null) return null
        return buildString {
            append("patient=$patientId")
            append("\napi=$apiBase")
            append("\norigin=$dashboardOrigin")
            authMethod?.let { append("\nauth=$it") }
        }
    }

    fun reconnect(): Boolean = runCatching {
        authenticate()
        true
    }.getOrDefault(false)

    private data class RegionalHosts(
        val regionalLoginUrl: String,
        val dashboardOrigin: String,
        val apiBase: String,
    )

    private data class HttpResult(
        val code: Int,
        val body: String,
        val url: String,
    ) {
        val isSuccessful: Boolean get() = code in 200..299
        val bodyPreview: String get() = body.take(300)
    }

    private fun executeRequest(url: String, preserveUrl: Boolean = false): HttpResult {
        val request = buildApiRequest(url, preserveUrl)
        var response = httpClient.newCall(request).execute()
        if (response.code == 401) {
            response.close()
            authenticate()
            response = httpClient.newCall(request).execute()
        }
        return response.use {
            HttpResult(
                code = it.code,
                body = it.body?.string().orEmpty(),
                url = it.request.url.toString(),
            )
        }
    }

    private fun buildApiRequest(url: String, preserveUrl: Boolean): Request {
        val requestBuilder = Request.Builder()
            .url(url)
            .get()
            .header("Accept", "application/json, text/plain, */*")
            .header("Content-Type", "application/json")
            .header("X-Requested-With", "XMLHttpRequest")
            .header("Origin", dashboardOrigin.orEmpty())
            .header("Referer", "${dashboardOrigin.orEmpty()}/")

        return if (preserveUrl) {
            requestBuilder.build().newBuilder().url(url).build()
        } else {
            requestBuilder.build()
        }
    }

    private fun apiRequest(url: String, preserveUrl: Boolean = false): JsonObject? {
        val result = executeRequest(url, preserveUrl)
        if (!result.isSuccessful) return null
        if (result.body.isBlank()) return null
        return runCatching { GlookoParser.parseJsonObject(result.body) }.getOrNull()
    }

    private fun formatIso(instant: Instant): String {
        return ISO_FORMATTER.format(instant.atOffset(ZoneOffset.UTC))
    }

    companion object {
        private const val GLOOKO_BASE_URL = "https://my.glooko.com"
        private const val GLOOKO_LOGIN_PATH = "/users/sign_in"
        private const val MAX_REDIRECTS = 10
        private val REDIRECT_CODES = 300..399
        private val JSON_MEDIA_TYPE = "application/json".toMediaType()
        private val DEVICE_ID = UUID.randomUUID().toString().replace("-", "").take(16)
        private val CSRF_TOKEN_REGEX = Regex("""name="csrf-token" content="([^"]+)"""")
        private val PATIENT_ID_PATTERNS = listOf(
            Regex("""window\.patient\s*=\s*"([^"]+)""""),
            Regex("""window\.patient\s*=\s*'([^']+)'"""),
            Regex("""window\.patientId\s*=\s*"([^"]+)""""),
            Regex("""window\.patientId\s*=\s*'([^']+)'"""),
            Regex(""""patient_glooko_code"\s*:\s*"([^"]+)""""),
            Regex(""""glooko_code"\s*:\s*"([^"]+)""""),
        )
        private val API_URL_REGEX = Regex("""apiUrl:\s*'([^']+)'""")
        private val ISO_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")

        fun defaultHttpClient(): OkHttpClient {
            val cookieJar = DomainMatchingCookieJar()

            return OkHttpClient.Builder()
                .cookieJar(cookieJar)
                .followRedirects(false)
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .addInterceptor { chain ->
                    val request = chain.request().newBuilder()
                        .header(
                            "User-Agent",
                            "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 " +
                                "(KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36",
                        )
                        .build()
                    chain.proceed(request)
                }
                .addInterceptor { chain ->
                    val original = chain.request()
                    val fixedUrl = original.url.toString()
                        .replace("%5B", "[")
                        .replace("%5D", "]")
                    if (fixedUrl == original.url.toString()) {
                        chain.proceed(original)
                    } else {
                        chain.proceed(original.newBuilder().url(fixedUrl).build())
                    }
                }
                .build()
        }
    }
}

private class DomainMatchingCookieJar : CookieJar {
    private val cookies = mutableListOf<Cookie>()

    override fun saveFromResponse(url: okhttp3.HttpUrl, newCookies: List<Cookie>) {
        if (newCookies.isEmpty()) return
        synchronized(cookies) {
            newCookies.forEach { newCookie ->
                cookies.removeAll {
                    it.name == newCookie.name &&
                        it.domain == newCookie.domain &&
                        it.path == newCookie.path
                }
                cookies.add(newCookie)
            }
        }
    }

    override fun loadForRequest(url: okhttp3.HttpUrl): List<Cookie> {
        synchronized(cookies) {
            return cookies.filter { it.matches(url) }
        }
    }
}

File: backend/g2g/glooko_client.py
"""Glooko HTTP client — ported from GlookoClient.kt."""

from __future__ import annotations

import re
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any
from urllib.parse import urljoin, urlparse, urlunparse

import requests

GLOOKO_BASE_URL = "https://my.glooko.com"
GLOOKO_LOGIN_PATH = "/users/sign_in"
DEVICE_ID = uuid.uuid4().hex[:16]

_UA = (
    "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36"
)


class GlookoAuthError(Exception):
    pass


@dataclass
class GlookoTestResult:
    patient_id: str
    api_base: str
    dashboard_origin: str
    device_count: int


@dataclass
class RegionalHosts:
    dashboard_origin: str
    api_base: str


def _decode_brackets(url: str) -> str:
    return url.replace("%5B", "[").replace("%5D", "]").replace("%5b", "[").replace("%5d", "]")


class BracketSession(requests.Session):
    """requests.Session that keeps literal [] in query strings."""

    def request(self, method, url, *args, **kwargs):  # type: ignore[override]
        return super().request(method, _decode_brackets(url), *args, **kwargs)


class GlookoClient:
    def __init__(
        self,
        email: str,
        password: str,
        session_timeout_minutes: int = 55,
        base_url: str = GLOOKO_BASE_URL,
        http_client: requests.Session | None = None,
    ):
        self.email = email
        self.password = password
        self.session_timeout_minutes = session_timeout_minutes
        self.base_url = base_url.rstrip("/")
        self.http = http_client or BracketSession()
        self.http.headers.update({"User-Agent": _UA})
        self.patient_id: str | None = None
        self.api_base: str | None = None
        self.dashboard_origin: str | None = None
        self.auth_method: str | None = None
        self.last_auth_time: datetime | None = None
        self.authenticate()

    def ensure_authenticated(self) -> None:
        if self.last_auth_time is None:
            self.authenticate()
            return
        elapsed = (datetime.now(timezone.utc) - self.last_auth_time).total_seconds() / 60
        if elapsed >= self.session_timeout_minutes:
            self.authenticate()

    def authenticate(self) -> None:
        self.patient_id = None
        self.api_base = None
        self.dashboard_origin = None
        self.auth_method = None

        regional = self._discover_regional_hosts()
        self.dashboard_origin = regional.dashboard_origin
        self.api_base = regional.api_base

        json_error = self._try_json_api_login(regional)
        if self.patient_id is not None:
            self.auth_method = "json-api"
            self.last_auth_time = datetime.now(timezone.utc)
            return

        self._try_web_form_login(regional, json_error)
        self.auth_method = "web-form"
        self.last_auth_time = datetime.now(timezone.utc)

    def _discover_regional_hosts(self) -> RegionalHosts:
        login_url = f"{self.base_url}{GLOOKO_LOGIN_PATH}?id=login_form&locale=en-GB"
        resp = self.http.get(login_url, allow_redirects=False, timeout=30)
        location = resp.headers.get("Location") or login_url
        if not location.startswith("http"):
            location = urljoin(self.base_url + "/", location.lstrip("/"))
        parsed = urlparse(location)
        origin = f"{parsed.scheme}://{parsed.netloc}"
        api_host = parsed.netloc.replace("my.glooko", "api.glooko")
        api_base = f"{parsed.scheme}://{api_host}"
        return RegionalHosts(dashboard_origin=origin, api_base=api_base)

    def _try_json_api_login(self, regional: RegionalHosts) -> str | None:
        url = f"{regional.api_base}/api/v2/users/sign_in"
        payload = {
            "userLogin": {"email": self.email, "password": self.password},
            "deviceInformation": {
                "applicationType": "logbook",
                "os": "android",
                "osVersion": "14",
                "device": "Android",
                "deviceManufacturer": "Google",
                "deviceModel": "Mobile",
                "serialNumber": DEVICE_ID,
                "deviceId": DEVICE_ID,
                "clinicalResearch": False,
                "applicationVersion": "6.1.3",
                "buildNumber": "0",
                "gitHash": "g4fbed2011b",
            },
        }
        headers = {
            "Accept": "application/json, text/plain, */*",
            "Content-Type": "application/json",
            "X-Requested-With": "XMLHttpRequest",
            "Origin": regional.dashboard_origin,
            "Referer": f"{regional.dashboard_origin}/",
        }
        try:
            resp = self.http.post(url, json=payload, headers=headers, timeout=30)
            if resp.status_code >= 400:
                return f"HTTP {resp.status_code}: {resp.text[:300]}"
            data = resp.json()
            code = (
                (data.get("userLogin") or {}).get("glookoCode")
                or ((data.get("user") or {}).get("userLogin") or {}).get("glookoCode")
                or data.get("glookoCode")
            )
            if not code:
                return "JSON login missing glookoCode"
            self.patient_id = str(code)
            self.api_base = regional.api_base
            return None
        except Exception as exc:
            return str(exc)

    def _try_web_form_login(self, regional: RegionalHosts, json_error: str | None) -> None:
        login_url = f"{regional.dashboard_origin}{GLOOKO_LOGIN_PATH}?id=login_form&locale=en-GB"
        page = self.http.get(login_url, allow_redirects=True, timeout=30)
        csrf_match = re.search(r'name="csrf-token"\s+content="([^"]+)"', page.text)
        if not csrf_match:
            raise GlookoAuthError(f"CSRF token not found. JSON login error: {json_error}")
        csrf = csrf_match.group(1)

        form = {
            "utf8": "✓",
            "authenticity_token": csrf,
            "user[email]": self.email,
            "user[password]": self.password,
            "commit": "Log In",
        }
        headers = {
            "Content-Type": "application/x-www-form-urlencoded",
            "Origin": regional.dashboard_origin,
            "Referer": login_url,
        }
        resp = self.http.post(
            f"{regional.dashboard_origin}{GLOOKO_LOGIN_PATH}",
            data=form,
            headers=headers,
            allow_redirects=False,
            timeout=30,
        )
        final_url = resp.url
        final_html = resp.text
        redirects = 0
        while resp.status_code in range(300, 400) and redirects < 10:
            loc = resp.headers.get("Location")
            if not loc:
                break
            next_url = urljoin(final_url, loc)
            headers["Referer"] = final_url
            resp = self.http.get(next_url, headers=headers, allow_redirects=False, timeout=30)
            final_url = next_url
            final_html = resp.text if resp.text else final_html
            redirects += 1

        if resp.status_code < 300:
            final_html = resp.text

        patient = self._extract_patient_id(final_html)
        if not patient and GLOOKO_LOGIN_PATH not in final_url:
            home = self.http.get(f"{regional.dashboard_origin}/", timeout=30)
            final_html = home.text
            patient = self._extract_patient_id(final_html)

        if not patient:
            raise GlookoAuthError(
                f"Web form login failed — patient ID not found. "
                f"JSON login error: {json_error}. Final URL: {final_url}"
            )

        self.patient_id = patient
        api_from_html = re.search(r"apiUrl:\s*'([^']+)'", final_html)
        if api_from_html:
            self.api_base = api_from_html.group(1).rstrip("/")
        else:
            parsed = urlparse(final_url)
            api_host = parsed.netloc.replace("my.glooko", "api.glooko")
            self.api_base = f"{parsed.scheme}://{api_host}" if parsed.scheme else regional.api_base
        self.dashboard_origin = regional.dashboard_origin

    @staticmethod
    def _extract_patient_id(html: str) -> str | None:
        patterns = [
            r'window\.patient\s*=\s*["\']([^"\']+)["\']',
            r'window\.patientId\s*=\s*["\']([^"\']+)["\']',
            r'"patient_glooko_code"\s*:\s*"([^"]+)"',
            r'"glooko_code"\s*:\s*"([^"]+)"',
        ]
        for pattern in patterns:
            m = re.search(pattern, html)
            if m:
                return m.group(1)
        return None

    def _format_ts(self, instant: datetime) -> str:
        if instant.tzinfo is None:
            instant = instant.replace(tzinfo=timezone.utc)
        return instant.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"

    def _api_get(self, path_and_query: str) -> dict[str, Any] | None:
        self.ensure_authenticated()
        assert self.api_base and self.dashboard_origin and self.patient_id
        url = _decode_brackets(f"{self.api_base}{path_and_query}")
        headers = {
            "Accept": "application/json, text/plain, */*",
            "Content-Type": "application/json",
            "X-Requested-With": "XMLHttpRequest",
            "Origin": self.dashboard_origin,
            "Referer": f"{self.dashboard_origin}/",
        }

        def do_get() -> requests.Response:
            return self.http.get(url, headers=headers, timeout=30)

        resp = do_get()
        if resp.status_code == 401:
            self.authenticate()
            resp = do_get()
        if resp.status_code >= 400 or not resp.text.strip():
            return None
        try:
            data = resp.json()
            return data if isinstance(data, dict) else None
        except Exception:
            return None

    def get_graph_data(self, start: datetime, end: datetime) -> dict[str, Any] | None:
        assert self.patient_id
        q = (
            f"/api/v3/graph/data?patient={self.patient_id}"
            f"&startDate={self._format_ts(start)}&endDate={self._format_ts(end)}"
            f"&locale=en-GB&insulinTooltips=true&filterBgReadings=true&splitByDay=false"
            f"&series[]=deliveredBolus"
        )
        return self._api_get(q)

    def get_statistics(self, start: datetime, end: datetime) -> dict[str, Any] | None:
        assert self.patient_id
        q = (
            f"/api/v3/graph/statistics/overall?patient={self.patient_id}"
            f"&startDate={self._format_ts(start)}&endDate={self._format_ts(end)}"
            f"&egv=false&includeInsulin=true&includeExercise=true"
            f"&dow=monday,tuesday,wednesday,thursday,friday,saturday,sunday"
            f"&includePumpModes=true"
        )
        return self._api_get(q)

    def get_device_settings(self) -> dict[str, Any] | None:
        assert self.patient_id
        return self._api_get(f"/api/v3/devices_and_settings?patient={self.patient_id}")

    def test_connection(self) -> GlookoTestResult:
        from g2g.glooko_parser import parse_devices

        data = self.get_device_settings()
        if data is None:
            raise RuntimeError("Failed to fetch device settings")
        devices = parse_devices(data)
        assert self.patient_id and self.api_base and self.dashboard_origin
        return GlookoTestResult(
            patient_id=self.patient_id,
            api_base=self.api_base,
            dashboard_origin=self.dashboard_origin,
            device_count=len(devices),
        )

    def session_summary(self) -> str | None:
        if not self.patient_id:
            return None
        return (
            f"patient={self.patient_id}\n"
            f"api={self.api_base}\n"
            f"origin={self.dashboard_origin}\n"
            f"auth={self.auth_method}"
        )
